src/linalg/impl_linalg.rs

```diff
@@ -823,3 +824,126 @@ mod blas_tests {
         assert!(blas_column_major_2d::<f32, _>(&m));
     }
 }
+
+#[allow(dead_code)]
+fn general_outer_to_dyn<Sa, Sb, F, T>(
+    a: &ArrayBase<Sa, IxDyn>,
+    b: &ArrayBase<Sb, IxDyn>,
+    mut f: F,
+) -> ArrayD<T>
+where
+    T: Copy,
+    Sa: Data<Elem = T>,
+    Sb: Data<Elem = T>,
+    F: FnMut(T, T) -> T,
+{
+    //Iterators on the shapes, compelted by 1s
+    let a_shape_iter = a.shape().iter().chain([1].iter().cycle());
+    let b_shape_iter = b.shape().iter().chain([1].iter().cycle());
+
+    let res_ndim = std::cmp::max(a.ndim(), b.ndim());
+    let res_dim: Vec<Ix> = a_shape_iter
+        .zip(b_shape_iter)
+        .take(res_ndim)
+        .map(|(x, y)| x * y)
+        .collect();
+
+    let mut res: ArrayD<MaybeUninit<T>> = ArrayBase::maybe_uninit(res_dim);
+    let res_chunks = res.exact_chunks_mut(b.shape());
+    Zip::from(res_chunks).and(a).apply(|res_chunk, &a_elem| {
+        Zip::from(b)
+            .apply_assign_into(res_chunk, |&b_elem| MaybeUninit::new(f(a_elem, b_elem)))
+    });
+    unsafe { res.assume_init() }
+}
+
+#[allow(dead_code, clippy::type_repetition_in_bounds)]
+fn kron_to_dyn<Sa, Sb, T>(a: &ArrayBase<Sa, IxDyn>, b: &ArrayBase<Sb, IxDyn>) -> Array<T, IxDyn>
+where
+    T: Copy,
+    Sa: Data<Elem = T>,
+    Sb: Data<Elem = T>,
+    T: crate::ScalarOperand + std::ops::Mul<Output = T>,
+{
+    general_outer_to_dyn(a, b, std::ops::Mul::mul)
+}
+
+#[allow(dead_code)]
+fn general_outer_same_size<Sa, I, Sb, F, T>(
+    a: &ArrayBase<Sa, I>,
+    b: &ArrayBase<Sb, I>,
+    mut f: F,
+) -> Array<T, I>
+where
+    T: Copy,
+    Sa: Data<Elem = T>,
+    Sb: Data<Elem = T>,
+    I: Dimension,
+    F: FnMut(T, T) -> T,
+{
+    let mut res_dim = a.raw_dim();
+    let mut res_dim_view = res_dim.as_array_view_mut();
+    res_dim_view *= &b.raw_dim().as_array_view();
```

Comment:
Obviously I kind of love this, that we can use array methods on dimensions 🙂.

The unsafe code looks good, we only need to prove that all elements are assigned to, and they will be if the exact chunks don't leave any uneven remainder. And that looks good to me, B's shape evenly divides the result's shape, right?

Reply:
Ah no, we also have to check the product. We need to use a saturating multiplication, otherwise we can overflow, and then the above doesn't hold?